docs/custom_promise_types/modules.md

```diff
@@ -0,0 +1,363 @@
+* Promise evaluation:
+  * `kept` - promise satisfied already, no change made
+  * `repaired` - promise not satisfied before, but fixed now
+    * The change should be explained in a `info` level log message
+  * `not_kept` - promse not satisfied before, and could not be fixed
+    * Should be explained by an `error` level log message
+* Teminate:
+  * `success` - Module succesfully terminated without errors
+  * `failure` - There were problems when trying to clean up / terminate
+    * Should be explained by a `critical` level log message
+
+Built-in CFEngine promises may have multiple outcomes when evaluated.
+For the sake of simplicity, custom promises may only have 1 outcome, logic should be as follows:
+
+1. If the promise was already satisfied and no changes were necessary, the promise was `kept`
+2. Otherwise, if there were issues and the module could not fix/satisfy all parts of the promise, it was `not_kept`.
+3. Otherwise, if some changes were made (successfully), the promise was `repaired`.
+
+### Protocol
+
+To implement a promise module from scratch, you will have to implement the promise module protocol.
+2 variants of the protocol is provided.
+1 JSON based and 1 line based.
+
+**Note that if you are using the library, as shown above, you don't have to care / learn about the following internals of the protocol.**
+
+### Protocol header
+
+When a promise module starts, the agent sends a protocol header (single line), followed by 2 newline characters.
+The promise module responds with a protocol header (single line), followed by 2 newline characters.
+
+The protocol header contains:
+
+* name of program
+* version of program
+* version of protocol
+* space separated list of features available (optional)
+
+The header has the same syntax regardless of protocol (It is used to determine protocol).
+
+#### Example headers:
+
+Request sent by cf-agent to promise module:
+
+```
+CFEngine 3.16.0 v1
+```
+
+Response, requesting line based protocol:
+
+```
+git_promises 0.0.1 v1 line_based
+```
+
+Response, requesting JSON based protocol:
+
+```
+git_promises 0.0.1 v1 json_based
+```
+
+After the initial exchange of protocol headers, the module responds to requests from the agent, using the agreed upon protocol.
+Requests and responses follow the same format and rules, and are commonly called messages.
+For each request, there should be exactly one response.
+
+### Protocol request sequence
+
+A promise module will receive 3 types of requests (operations):
+
+1. `validate_promise` - Validate the names and types of promise attributes
+2. `evaluate_promise` - Perform changes to the system
+3. `terminate` - Shut down the module
+
+The module should not make any assumptions about the order of validation or evaluation.
+Each promise might be validated and evaluated, 0, 1, or more times, in any order.
+In general, the promise module should not maintain any state between the individual requests.
```

Comment:
I would expect the module to potentially cache some things. So I think we don't have to mention explicitly that it should not maintain any state.

Reply:
Changed wording.